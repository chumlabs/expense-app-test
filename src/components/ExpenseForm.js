// IMPORTS
// - libraries
import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize'; // needed for react-dates
import { SingleDatePicker } from 'react-dates';
import { Redirect } from 'react-router-dom';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toFixed(2).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = date => {
    // note: date is a moment object
    // if (date) {
    //   this.setState(() => ({ createdAt: date }));
    // }
    this.setState(() => ({ createdAt: date }));
    // console.log(`change to date: `, date);
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
    // optional validation
    // if (!this.state.createdAt) {
    //   this.setState(() => ({ error: 'Please enter a valid date' }));
    // } else {
    //   this.setState(() => ({ error: '' }));
    // }
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description) {
      this.setState(() => ({ error: 'Please enter a description' }));
    } else if (!this.state.amount) {
      this.setState(() => ({ error: 'Please enter an amount' }));
    } else if (!this.state.createdAt) {
      this.setState(() => ({ error: 'Please enter a valid date' }));
    } else {
      this.setState(() => ({ error: '' }));
      // console.log('submitted!');
      // use onSubmit prop
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form onSubmit={this.onSubmit} className="form-group">
        <div className="form-group__error">
          {this.state.error ? <p className="error-text">{this.state.error}</p> : null}
        </div>
        <input
          className="form-group__text-input"
          type="text"
          placeholder="Description (required)"
          value={this.state.description}
          onChange={this.onDescriptionChange}
          autoFocus
          required
        />

        <input
          className="form-group__text-input"
          type="text"
          placeholder="Amount (required)"
          value={this.state.amount}
          onChange={this.onAmountChange}
          required
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />

        <textarea
          className="form-group__textarea"
          name="notes"
          id=""
          cols="40"
          rows="3"
          placeholder="Additional Notes (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        />

        <div>
          <button className="button form-group__submit-button">
            {this.props.actionText}
          </button>
        </div>
      </form>
    );
  }
}
