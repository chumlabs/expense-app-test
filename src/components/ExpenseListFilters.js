// IMPORTS
// - libraries
import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
// - components
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filterActions';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    // console.log(e.target.value);
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="filter-group">
          <div className="filter-group__item">
            <input
              type="text"
              className="filter-group__text-input"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Enter text to search"
            />
          </div>
          <div className="filter-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
              startDateId={'start'}
              endDateId={'end'}
            />
          </div>
          <div className="filter-group__item">
            Sort By:&nbsp;
            <select
              name="sort"
              id="sort"
              className="filter-group__select-box"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">date</option>
              <option value="amount">amount</option>
            </select>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: date => dispatch(setStartDate(date)),
  setEndDate: date => dispatch(setEndDate(date)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
