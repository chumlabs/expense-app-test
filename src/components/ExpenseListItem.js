// IMPORTS
// - libraries
import React from 'react';
// import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
// - components
// import { removeExpense } from '../actions/expenseActions';

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`} className="list-item">
      <div>
        <h3 className="list-item__title">{description} </h3>
        <span className="list-item__subtitle">
          {moment(createdAt).format('MMM Do, YYYY')}
        </span>
      </div>
      <div>
        <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
      </div>
    </Link>
  </div>
);

export default ExpenseListItem;
