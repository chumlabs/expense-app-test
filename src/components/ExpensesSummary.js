import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotalAmount }) => {
  const expenseTense = expensesCount === 1 ? 'expense' : 'expenses';
  const formattedAmount = numeral(expensesTotalAmount / 100).format('$0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Summary</h1>
        <p className="page-header__summary">
          Showing <span>{expensesCount}</span> {expenseTense} totalling{' '}
          <span>{formattedAmount}</span>
        </p>
        <div className="page-header__actions">
          <Link to="/create" className="button">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  const filteredExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expensesCount: filteredExpenses.length,
    expensesTotalAmount: selectExpensesTotal(filteredExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
