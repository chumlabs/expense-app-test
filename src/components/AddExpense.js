// IMPORTS
// - libraries
import React, { Component } from 'react';
import { connect } from 'react-redux';
// - components
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenseActions';

export class AddExpense extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} actionText="Save Expense" />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(null, mapDispatchToProps)(AddExpense);
