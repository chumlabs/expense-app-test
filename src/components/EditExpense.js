// IMPORTS
// - libraries
import React, { Component } from 'react';
import { connect } from 'react-redux';
// - components
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenseActions';

export class EditExpense extends Component {
  // { dispatch, expense, match, history }
  onSubmit = exp => {
    // console.log('expense updated', exp);
    this.props.startEditExpense(this.props.expense.id, exp);
    this.props.history.push('/dashboard');
  };

  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
            actionText="Update Expense"
          />
          <button onClick={this.onClick} className="button button--secondary">
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  expense: state.expenses.find(exp => exp.id === ownProps.match.params.id)
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  startEditExpense: (id, exp) => dispatch(startEditExpense(id, exp)),
  startRemoveExpense: idObj => dispatch(startRemoveExpense(idObj))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
