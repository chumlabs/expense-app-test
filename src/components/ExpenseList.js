// IMPORTS
// -libraries
import React from 'react';
import { connect } from 'react-redux';
// - components
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// presentational component
export const ExpenseList = ({ expenses }) => (
  <div className="content-container">
    <div className="list">
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
      <div className="list-body">
        {expenses.length !== 0 ? (
          expenses.map(exp => (
            <ExpenseListItem {...exp} key={exp.id} className="list__item" />
          ))
        ) : (
          <div className="list-item list-item--message">
            Create an expense or change the filters to see additional expenses
          </div>
        )}
      </div>
    </div>
  </div>
);

// mapStateToProps func (convention)
const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

// connect funct (HOC component)
// - composes presentational component with state
export default connect(mapStateToProps)(ExpenseList);
