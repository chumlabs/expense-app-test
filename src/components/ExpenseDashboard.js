// IMPORTS
// - libraries
import React, { Component } from 'react';
// - components
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
