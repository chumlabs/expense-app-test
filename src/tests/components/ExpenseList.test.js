import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenseData'; // mock data

test('should render the ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList {...expenses[0]} />);

  expect(wrapper).toMatchSnapshot();
});

test('should render the ExpenseList when no there are no expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);

  expect(wrapper).toMatchSnapshot();
});
