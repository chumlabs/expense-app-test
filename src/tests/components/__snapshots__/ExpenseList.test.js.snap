// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the ExpenseList when no there are no expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list"
  >
    <div
      className="list-header"
    >
      <div
        className="show-for-mobile"
      >
        Expenses
      </div>
      <div
        className="show-for-desktop"
      >
        Expense
      </div>
      <div
        className="show-for-desktop"
      >
        Amount
      </div>
    </div>
    <div
      className="list-body"
    >
      <div
        className="list-item list-item--message"
      >
        Create an expense or change the filters to see additional expenses
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the ExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list"
  >
    <div
      className="list-header"
    >
      <div
        className="show-for-mobile"
      >
        Expenses
      </div>
      <div
        className="show-for-desktop"
      >
        Expense
      </div>
      <div
        className="show-for-desktop"
      >
        Amount
      </div>
    </div>
    <div
      className="list-body"
    >
      <ExpenseListItem
        amount={700}
        className="list__item"
        createdAt={1491062400000}
        description="item 1 - abc"
        id="1"
        key="1"
        note="lowest amount, middle date"
      />
      <ExpenseListItem
        amount={2000}
        className="list__item"
        createdAt={1488387600000}
        description="item 2 - bcd"
        id="2"
        key="2"
        note="highest amount, earliest date"
      />
      <ExpenseListItem
        amount={1100}
        className="list__item"
        createdAt={1493654400000}
        description="item 3 - cdf"
        id="3"
        key="3"
        note="middle amount,  latest date"
      />
    </div>
  </div>
</div>
`;
